AuthenticatorTests/TokenListViewControllerTest.swift

```diff
@@ -0,0 +1,92 @@
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+//  SOFTWARE.
+//
+
+import XCTest
+import OneTimePassword
+@testable import Authenticator
+
+class TokenListViewControllerTest: XCTestCase {
+
+    func emptyListViewModel() -> TokenList.ViewModel {
+        return mockList([]).viewModel
+    }
+
+    // Test that inserting a new token will produce the expected changes to the table view.
+    func testTokenListInsertsNewToken() {
+        // Set up a view controller with a mock table view.
+        let initialViewModel = emptyListViewModel()
+        let controller = TokenListViewController(viewModel: initialViewModel, dispatchAction: { _ in })
+        let tableView = MockTableView()
+        controller.tableView = tableView
+
+        // Update the view controller.
+        let updatedViewModel = mockList([
+            ("Service", "email@example.com"),
+        ]).viewModel
+        controller.updateWithViewModel(updatedViewModel)
+
+        // Check the table view.
+        let expectedChanges: [MockTableView.ChangeType] = [
+            .BeginUpdates,
+            .Insert(indexPath: NSIndexPath(forRow: 0, inSection: 0)),
+            .EndUpdates,
+        ]
+        XCTAssertEqual(tableView.changes, expectedChanges)
+    }
+
+    // Test that updating an existing token will produce the expected changes to the table view.
+    func testUpdatesExistingToken() {
+        // Set up a view controller with a mock table view.
+        let displayTime = DisplayTime(date: NSDate())
+        let initialPersistentToken = mockPersistentToken(name: "account@example.com", issuer: "Issuer")
+        let initialTokenList = TokenList(persistentTokens: [initialPersistentToken], displayTime: displayTime)
+        let controller = TokenListViewController(viewModel: initialTokenList.viewModel, dispatchAction: { _ in })
+        let tableView = MockTableView()
+        controller.tableView = tableView
+
+        // Update the view controller.
+        let updatedPersistentToken = initialPersistentToken.updated(with: mockToken(name: "name", issuer: "issuer"))
+        let updatedTokenList = TokenList(persistentTokens: [updatedPersistentToken], displayTime: displayTime)
+        controller.updateWithViewModel(updatedTokenList.viewModel)
+
+        // Check the table view.
+        // Updates to existing rows should be applied directly to the cells, without changing the table view.
+        // TODO: Test for direct updates to the cell via `updateWithRowModel`.
+        let expectedChanges: [MockTableView.ChangeType] = []
```

Comment:
We could potentially ask the controller for a table view cell here and check that the cell has the updated token's values.